### PR TITLE
[Mig] product_mandatory_products into 15.0

### PR DESCRIPTION
**Pull request by nikunjantala**

### CODE REVIEW

**Comment by jans23** [comment]
@dsolanki-initos Please review.

**Review by dsolanki-initos — COMMENTED** [review]
Code review and functional test!

I have compared functionality with v12.0 and found that we get a warning when we choose more than one mandatory product while checkout and in v15.0 we can not redirect to checkout by a single click as per v12.0 functionality I don't know if it is intentionally changed or by mistake.

Commented code can be removed in both JS files and python files.

**Comment by jans23** [comment]
> we can not redirect to checkout by a single click as per v12.0 functionality

Go to Settings and disable "Stay on page after adding to cart".

**Comment by dsolanki-initos** [comment]
I've checked with disabling "Stay on page after adding to cart" and still not getting the same as v12.0.
For your reference, I've attached a video let me know if I am missing something here

[mandatory_check.webm](https://user-images.githubusercontent.com/50451593/189103754-4843d294-4792-4e53-8349-d67d22d948a6.webm)


> > we can not redirect to checkout by a single click as per v12.0 functionality
> 
> Go to Settings and disable "Stay on page after adding to cart".

**Comment by jans23** [comment]
The video looks good to me. What is the exact issue you experience?

**Comment by dsolanki-initos** [comment]
> The video looks good to me. What is the exact issue you experience?

So we have "proceed to checkout" button click twice for the checkout page in v15.0?

**Comment by jans23** [comment]
In your video I see the first overlay windows is for mandatory products, pressing  "proceed to checkout" brings you to the 2nd overlay window which is for optional products, pressing  "proceed to checkout" brings you to the checkout. Nothing wrong with it.

**Comment by dsolanki-initos** [comment]
> In your video I see the first overlay windows is for mandatory products, pressing "proceed to checkout" brings you to the 2nd overlay window which is for optional products, pressing "proceed to checkout" brings you to the checkout. Nothing wrong with it.

Okay

File: product_mandatory_products/controller/main.py [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+                    product_and_options
+                )
+            res = super(WebsiteSaleExtend, self).cart_options_update_json(
+                product_and_options, goto_shop=goto_shop, lang=lang, **kwargs
+            )
+            self.KW = []
+            return res
+        else:
+            self.KW.clear()
+            self.KW.extend(json.loads(product_and_options))
```
Comment (manengel) [inline product_mandatory_products/controller/main.py:115]: I have some question about this....is the content of KW subject to "surviving" different route calls?

File: product_mandatory_products/models/__init__.py [hunk 2]
```diff
@@ -0,0 +1,4 @@
+from . import product
+
+# Not in use at the moment
+# from . import sale_order
```
Comment (manengel) [inline product_mandatory_products/models/__init__.py]: Well...when does it get in discussion if that code is used again?

File: product_mandatory_products/models/sale_order.py [hunk 3]
```diff
@@ -0,0 +1,59 @@
+from odoo import _, models
```
Comment (manengel) [inline product_mandatory_products/models/sale_order.py]: See comment regarding __init__.py in the same folder.

File: product_mandatory_products/static/src/js/website_sale_options.js [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+  var publicWidget = require("web.public.widget");
+  var OptionalProductsModal = require("sale_product_configurator.OptionalProductsModal");
+  require("website_sale_options.website_sale");
+
+  publicWidget.registry.WebsiteSale.include({
+    //    _onProductReady: function () {
+    //        debugger;
+    //        var res =  this._super.apply(this, arguments);
+    //        return res
+    //    },
```
Comment (manengel) [inline product_mandatory_products/static/src/js/website_sale_options.js]: Do you think there might be a future use for this commented code or can it be removed?

File: product_mandatory_products/views/product_template_view.xml [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+    <record id="mandatory_product_ids_product_template_form_view" model="ir.ui.view">
+        <field name="name">mandatory.product.ids.template.product.form</field>
+        <field name="model">product.template</field>
+        <field name="inherit_id" ref="product.product_template_form_view" />
+        <field name="arch" type="xml">
+            <xpath expr="//group[@name='upsell']" position="inside">
+                <field
+          name="mandatory_product_ids"
+          widget="many2many_tags"
+          domain="[('id', '!=', active_id)]"
```
Comment (manengel) [inline product_mandatory_products/views/product_template_view.xml:13]: I am not sure what the intended purpose of that domain is. From my understanding, it should lead to only relevant product_ids be displayed. But aren't things with active = False not hidden by Odoo anyways, normally?

File: product_mandatory_products/views/template.xml [hunk 6]
```diff
@@ -0,0 +1,376 @@
...
+          t-attf-href="#"
+          class="btn btn-primary float_left js_add_cart_json d-none d-md-inline-block"
+          aria-label="Add one"
+          title="Add one"
+        >
+                    <i class="fa fa-plus" />
+                </button>
+            </div>
+        </div>
+    </template>
```
Comment (manengel) [inline product_mandatory_products/views/template.xml]: Formatting looks partially very weird.

File: product_mandatory_products/views/template.xml [hunk 7]
```diff
@@ -0,0 +1,376 @@
...
+            >This combination does not exist.</p>
+                    </div>
+                    <div class="col-lg-1 d-inline-block" />
+                    <div class="col-lg-5 d-inline-block align-top text-left">
+                        <img
+              t-if="product_variant"
+              t-att-src="'/web/image/product.product/%s/image_1024' % product_variant.id"
+              class="d-block product_detail_img"
+              alt="Product Image"
+            />
```
Comment (manengel) [inline product_mandatory_products/views/template.xml]: Here, too.

**Review by manengel — COMMENTED** [review]
Left some feedback, but did not test yet. Please check my comments.

**Comment by manengel** [comment]
@nikunjantala @jan23 I do not know if it is intended like this. In my test scenario I could also link two or none instead of just one mandatory product. I could select none of them, click on "Weiter Einkaufen" and would have Product A in my shopping cart without choosing either Product B or Product C, which are linked to it as mandatory. On the other hand, I could also "resist" the prompt to choose exactly one mandatory product by choosing both B and C!

File: product_mandatory_products/views/template.xml (hunk 6 above)
Comment (dsolanki-initos) [reply product_mandatory_products/views/template.xml]: @manengel formatted the code.
@jans23 Shall I remove the commented code or I ask nikunj related to it?

**Comment by jans23** [comment]
Correct behaviour is that exactly one product has to be chosen. Not zero, not more than one but exactly one.

File: product_mandatory_products/views/template.xml (hunk 6 above)
Comment (jans23) [reply product_mandatory_products/views/template.xml]: Feel free to remove commented code.

Comment (dsolanki-initos) [reply product_mandatory_products/views/template.xml]: @jans23 I've removed the extra commented code.

**Review by manengel — CHANGES_REQUESTED** [review]
I did test it yet again. So basically now it prevents one from selecting none of the mandatory products. But it is still possible to select more than one!

I think this is contributed to the has_mandatory variable and the way it is used, but this is only judgement based on a quick look.

**Comment by dsolanki-initos** [comment]
@jans23, I've tested this module and also test it v12 on shoptest1,
Case on shoptest1 v12: When we select two mandatory products from the popup so it raises a warning that is"Please choose exactly one mandatory product."

Case in this PR v15: When we select two mandatory products from the popup so it will not raise any warning.

While testing I found a bug in this module of v15 if we click on the "continue shopping" button of the popup so it will add the product to add to cart without adding the mandatory product and it's not raising any warning.

**Comment by alihashaam** [comment]
[](url)@jans23 @nikunjantala @dsolanki-initos 

The code is reviewed and tested. During testing I encountered an error. LIke when we click the 'add to cart' button for Mandatory Items (see the First screenshot), we get a JS error (second screenshot). It does not interrupts our backend functionality. Apparently, it was caused by JS code in nitrokey_nitropad. I have applied a quick fix in 

https://git.initos.com/nitrokey/odoo-modules/-/merge_requests/563

So as soon as the above mentioned MR is merged. We can close this PR (@jans23 )

Selection_032


Selection_033

**Pull request merged**